akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -269,6 +270,57 @@ object GraphStages {
   def terminationWatcher[T]: GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]] =
     TerminationWatcher.asInstanceOf[GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]]]
 
+  private class FlowMonitorImpl[T] extends FlowMonitor[T] {
+    private val stateRef = new AtomicReference[StreamState[T]](Initialized)
+    def setState(state: StreamState[T]): Unit = stateRef.set(state)
+
+    override def state = stateRef.get()
+  }
```

Comment:
hm, that can be parried by the onPush checking that if the grabbed elem is instance of StreamState and if it is then wrap it in a Received().